compiler/src/iree/compiler/Dialect/Flow/Transforms/HoistEncodingOps.cpp

```diff
@@ -0,0 +1,242 @@
+                                      linalg::BroadcastOp broadcastOp) {
+  if (!broadcastOp->hasOneUse()) {
+    return failure();
+  }
+  RankedTensorType encodedType = encodingOp.getResultType();
+
+  // Create new encoding and set encoding on broadcast input.
+  AffineMap bcastMap = getBcastMapOrIdentity(rewriter, encodedType);
+  bcastMap = bcastMap.dropResults(broadcastOp.getDimensions());
+  Value input = broadcastOp.getInput();
+  auto encoding = cast<IREE::Encoding::EncodingAttr>(encodedType.getEncoding());
+  auto newEncoding = encoding.clone(bcastMap);
+  auto inputType = cast<RankedTensorType>(input.getType());
+  auto resType = RankedTensorType::get(inputType.getShape(),
+                                       inputType.getElementType(), newEncoding);
+  Location loc = broadcastOp->getLoc();
+  IREE::Encoding::SetEncodingOp newSetEncoding =
+      rewriter.create<IREE::Encoding::SetEncodingOp>(loc, resType, input);
+
+  // Create encoded broadcast op.
+  SmallVector<OpFoldResult> mixedSizes =
+      tensor::getMixedSizes(rewriter, loc, encodingOp.getSource());
+  Value encodedBCastInit = rewriter.create<tensor::EmptyOp>(
+      loc, mixedSizes, encodedType.getElementType(), encoding);
+  SmallVector<Value> encodedBCastOperands = {newSetEncoding.getResult(),
+                                             encodedBCastInit};
+  auto encodedBroadcast = clone(
+      rewriter, broadcastOp, encodingOp.getResultType(), encodedBCastOperands);
+
+  rewriter.replaceOp(encodingOp, encodedBroadcast);
+  return success();
+}
+
+static LogicalResult
+bubbleUpSetEncodingThroughGenericOp(RewriterBase &rewriter,
+                                    IREE::Encoding::SetEncodingOp encodingOp,
+                                    linalg::GenericOp genericOp) {
+  if (!genericOp->hasOneUse()) {
+    return failure();
+  }
+  if (genericOp.getNumDpsInits() != 1) {
+    return failure();
+  }
+  if (genericOp.getNumReductionLoops() != 0) {
+    return failure();
+  }
+  AffineMap outputMap =
+      genericOp.getMatchingIndexingMap(genericOp.getDpsInitOperand(0));
+  if (!outputMap.isIdentity()) {
+    return rewriter.notifyMatchFailure(genericOp, "output map not identity");
+  }
+
+  RankedTensorType encodedType = encodingOp.getResultType();
+  AffineMap bcastMap = getBcastMapOrIdentity(rewriter, encodedType);
+  if (outputMap.getNumDims() != bcastMap.getNumDims()) {
+    return rewriter.notifyMatchFailure(
+        genericOp, "output map dims do not match bcast_map dims");
+  }
+
+  // Set encodings on each input
+  Location loc = genericOp->getLoc();
+  SmallVector<Value> encodedOperands;
+  auto encoding = cast<IREE::Encoding::EncodingAttr>(encodedType.getEncoding());
+  for (OpOperand *operand : genericOp.getDpsInputOperands()) {
+    // Compute the new bcastMap from the operand's indexing map.
+    AffineMap operandMap = genericOp.getMatchingIndexingMap(operand);
+    AffineMap newBcastMap = operandMap.compose(bcastMap);
```

Comment:
If there are chained broadcasts, then we need to handle that case because we need to move the set_encoding op out of the dispatch. Once the set_encoding op is outside the dispatch, though, it will not be bubbled anymore. I was not considering the `broadcast->broadcast->set_encoding` when I implemented this because I did not think it would show up, so I sort of incidentally supported that case. I can either add a test to confirm that it is working (since in general we would need to support that case anyway), or I can add a check that the bcast_map is identity before bubbling. What do you think?

Reply:
Adding a check that the bcast_map is identity sounds better to me. You can leave a note saying that it is doable, but it is not common in IREE default pipeline.